### PR TITLE
fix #1831, Services exits on bind failure

### LINKED ISSUE #1381: test on travis out of memory

### LINKED ISSUE #1831: Services daemon swallows bind failure
so that a Services process is launched even if the address is already bound

### PR DESCRIPTION
**Pull request by manuzhang**

### CODE REVIEW

File: docs/faq.md [hunk 1]
```diff
@@ -33,5 +33,9 @@ Please check whether you are doing blocking jobs (e.g. sleep, IO) in your task.
 
 Generally, we recommend use the default `share-thread-pool-dispatcher` which has better performance and only turn to the `single-thread-dispatcher` when you have to.
 
+##### Why can't I open Dashboard even if the Services process has been launched successfully ?
+
+By default, our Services process binds to a local **IPv6 port**. It's possible that another process on your system has already taken up the same **IPv4 port**. You may check by `lsof -i -P | grep -i "Listen"` if your system is Unix/Linux.
```
Comment (clockfly) [inline docs/faq.md]: Can we disallow binding to IPV6?

File: services/jvm/src/main/scala/io/gearpump/services/RestServices.scala [hunk 2]
```diff
@@ -108,7 +105,11 @@ object RestServices {
 
     implicit val materializer = ActorMaterializer()
 
-    Http().bindAndHandle(Route.handlerFlow(services.routes), host, port)
+    Http().bindAndHandle(Route.handlerFlow(services.routes), host, port) onFailure {
+      case t: Throwable =>
+        System.err.println(t)
+        System.exit(-1)
```
Comment (clockfly) [inline services/jvm/src/main/scala/io/gearpump/services/RestServices.scala]: System.exit is too disruptive. It is dangerous to use it in YARN. Can you throw an exception instead?

Comment (manuzhang) [inline services/jvm/src/main/scala/io/gearpump/services/RestServices.scala]: exception cannot stop the process from running

File: docs/faq.md (hunk 1 above)
Comment (manuzhang) [inline docs/faq.md]: well, it's possible another process takes the same IPv6 port.

File: services/jvm/src/main/scala/io/gearpump/services/RestServices.scala (hunk 2 above)
Comment (manuzhang) [inline services/jvm/src/main/scala/io/gearpump/services/RestServices.scala]: the exception is thrown in future and doesn't stop the process from running

**Comment by clockfly** [comment]
System.exit(-1) is not acceptable. 

It will stop the UT, and also break the YARN module.

File: docs/faq.md (hunk 1 above)
Comment (clockfly) [inline docs/faq.md]: You mean 
When:
1. When another process take the ipv6 port 8090.
2. UI successfully bind to ipv4 port 8090.

the user cannot open dashboard of the ipv4:8090?

Comment (manuzhang) [inline docs/faq.md]: yes

**Comment by clockfly** [comment]
@manuzhang 


Are you going to revise the patch?

**Comment by manuzhang** [comment]
oh, yes

**Comment by manuzhang** [comment]
@clockfly updated

**Comment by clockfly** [comment]
+1

**Pull request merged**